lib/ansible/modules/windows/win_domain_object_info.ps1

```diff
@@ -0,0 +1,96 @@
+#!powershell
+
+# Copyright: (c) 2019, Shachaf Goldstein <shachaf.gold@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+Set-StrictMode -Version 2.0
+
+$spec = @{
+    options = @{
+        filter = @{ type = "str"; required = $false; }
+        identity = @{ type = "str"; required = $false}
+        include_deleted_objects = @{ type = "bool"; required = $false; default = $false }
+        json_depth = @{ type = "int"; required = $false; default = 3 }
+        properties = @{ type = "list"; required = $false; default = @("*"); elements = "str" }
+        server = @{ type = "str"; required = $false; aliases = @("domain_server") }
+        ldap_filter = @{ type = "str"; required = $false; }
+        search_scope = @{ type = "str"; choices = @("Base","OneLevel","Subtree"); required = $false; default = "Subtree" }
+        search_base = @{ type = "str"; required = $false; }
+        username = @{ type = "str"; required = $false; aliases = @("domain_username") }
+        password = @{ type = "str"; required = $false; aliases = @("domain_password") }
+    }
+    required_together = @(,@("username","password"))
+    required_one_of = @(,@("filter","identity","ldapFilter"))
+    supports_check_mode = $true
+}
+
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
+
+$filter = $module.Params.filter
+$identity = $module.Params.identity
+$include_deleted_objects = $module.Params.include_deleted_objects
+$properties = $module.Params.properties
+$server = $module.Params.server
+$ldap_filter = $module.Params.ldap_filter
+$search_scope = $module.Params.search_scope
+$search_base = $module.Params.search_base
+$username = $module.Params.username
+$password = ConvertTo-SecureString -AsPlainText -Force -String $module.Params.password
+$json_depth = $module.Params.json_depth
+
+$module.result = @{
+    objects = @()
+}
+
+$args_hashtable = @{}
+
+if ($null -ne $filter)
+{
+    $args_hashtable.Filter = $filter
+}
+
+if ($null -ne $identity)
+{
+    $args_hashtable.Identity = $identity
+}
+
+if ($null -ne $includeDeletedObjects)
+{
+    $args_hashtable.IncludeDeletedObjects = $include_deleted_objects
+}
+
+if ($null -ne $properties)
+{
+    $args_hashtable.Properties = $properties
+}
+
+if ($null -ne $server)
+{
+    $args_hashtable.Server = $server
+}
+
+if ($null -ne $ldapFilter)
+{
+    $args_hashtable.LDAPFilter = $ldap_filter
+}
+
+if ($null -ne $searchScope)
+{
+    $args_hashtable.SearchScope = $search_scope
+}
+
+if ($null -ne $searchBase)
+{
+    $args_hashtable.SearchBase = $search_base
+}
+
+if ($null -ne $username -and $null -ne $password)
+{
+    $args_hashtable.Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $username, $password
+}
+
+$result.objects = Get-ADObject @args_hashtable | ForEach-Object {$_ | ConvertTo-Json -Depth $json_depth -Compress}
```

Comment:
This is a bit of a tricky one, right now this approach has a few issues

* The output of `Get-ADObject` is an actual [ADObject](https://docs.microsoft.com/en-us/dotnet/api/microsoft.activedirectory.management.adobject?view=activedirectory-management-10.0) class
* This class has some metadata like `PropertyNames`, `AddedProperties`, `ObjectGuid`, which are either doubled up on the properties returned or just general metadata not useful for the end user
* This is 1 of the issues where the `json_depth` option came in handy but if we don't return them (because it's not needed) we remove 1 need for `json_depth`
* We should be filtering the result with the properties we requested (or if `properties: '*'` was set, the `PropertyNames` returned).
* Some of the properties return either large amounts of information (`nTSecurityDescriptor`), hard to parse (any DateTime or Octet Strings), or extra useless metadata (SID)

To solve these issues I think the following code should be:

```
Function ConvertTo-OutputValue {
    [CmdletBinding()]
    Param (
        [Parameter(Mandatory=$true)]
        [AllowNull()]
        [Object]
        $InputObject
    )

    if ($InputObject -is [System.Security.Principal.SecurityIdentifier]) {
        # Syntax: SID - Only serialize the SID as a string and not the other metadata properties.
        $InputObject.Value
    } elseif ($InputObject -is [Byte[]]) {
        # Syntax: Octet String - By default will serialize as a list of decimal values per byte, instead return a
        # Base64 string as Ansible can easily parse that.
        [System.Convert]::ToBase64String($InputObject)
    } elseif ($InputObject -is [DateTime]) {
        # Syntax: UTC Coded Time - .NET DateTimes serialized as in the form "Date(FILETIME)" which isn't easily
        # parsable by Ansible, instead return as an ISO 8601 string.
        $InputObject.ToString("o")
    } elseif ($InputObject -is [System.Security.AccessControl.ObjectSecurity]) {
        # Complex object which isn't easily serializable. Instead we should just return the SDDL string. If a user
        # needs to parse this then they really need to reprocess the SDDL string and process their results on another
        # win_shell task.
        $InputObject.GetSecurityDescriptorSddlForm(([System.Security.AccessControl.AccessControlSections]::All))
    } else {
        # Syntax: (All Others) - The default serialization handling of other syntaxes are fine, don't do anything.
        $InputObject
    }
}

foreach ($ad_object in (Get-ADObject @args_hashtable)) {
    if ('*' -in $properties) {
        # The user specified all properties, return the attributes that were retrieved from LDAP.
        $property_names = $ad_object.PropertyNames
    } else {
        $property_names = $properties
    }

    # Now process each property to an easy to represent string
    $filtered_object = @{}
    foreach ($name in $property_names) {
        $value = $ad_object.$name
        if ($value -is [Microsoft.ActiveDirectory.Management.ADPropertyValueCollection]) {
            $value = foreach ($v in $value) {
                ConvertTo-OutputValue -InputObject $v
            }
        } else {
            $value = ConvertTo-OutputValue -InputObject $value
        }
        $filtered_object.$name = $value
    }
    $module.Result.objects.Add($filtered_object)
}
```

Doing this;

* Ensures the `objects` field is always a list, empty for no objects else list of dict(s)
    * The behaviour before this change resulted in a few different output types which is not ideal (no object == null, 1 object == json dict as string, 2+ objects == json list with dicts as string)
* Removes the need for serializing and deserializing in JSON as we are not longer returning any complex objected that have nested properties (Security Descriptor or ADObject .NET Properties)
    * Because of this we no longer need `json_depth`
* Makes the values easier to parse in Ansible
    * Bytes are now a base64 string which can easily be decoded in Ansible with `{{ res.prop | b64decode }}`
    * A Security Descriptor is now an SDDL string which is the proper serialized form. You still require another Windows task to properly process this but we no longer loose any info and the value is a string not a extremely nested and complex object
    * Datetimes are now an ISO 8601 string which can be parsed in Python but I'm not 100% sure if there's a builtin filter in Ansible, in any case it's easier than the `Datetime(FILETIME)` format that was being returned before
    * SIDs now just return the SID as a string and not the metadata around that like bytes required

Ultimately here is the output before the above change (json_depth = 1)

```
ok: [dc01] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "filter": "sAMAccountName -eq 'vagrant'",
            "identity": null,
            "include_deleted_objects": false,
            "ldap_filter": null,
            "password": null,
            "properties": [
                "*"
            ],
            "search_base": null,
            "search_scope": null,
            "server": null,
            "username": null
        }
    },
    "objects": "{\"DistinguishedName\":\"CN=vagrant,CN=Users,DC=domain,DC=local\",\"Name\":\"vagrant\",\"ObjectClass\":\"user\",\"ObjectGuid\":\"8162ff17-3156-4aec-8c27-9a193de34a10\",\"PropertyNames\":[\"accountExpires\",\"adminCount\",\"audio\",\"badPasswordTime\",\"badPwdCount\",\"CanonicalName\",\"CN\",\"codePage\",\"countryCode\",\"Created\",\"createTimeStamp\",\"Deleted\",\"Description\",\"DisplayName\",\"DistinguishedName\",\"dSCorePropagationData\",\"instanceType\",\"isDeleted\",\"LastKnownParent\",\"lastLogoff\",\"lastLogon\",\"lastLogonTimestamp\",\"logonCount\",\"logonHours\",\"memberOf\",\"Modified\",\"modifyTimeStamp\",\"Name\",\"nTSecurityDescriptor\",\"ObjectCategory\",\"ObjectClass\",\"ObjectGUID\",\"objectSid\",\"primaryGroupID\",\"ProtectedFromAccidentalDeletion\",\"pwdLastSet\",\"sAMAccountName\",\"sAMAccountType\"\"sDRightsEffective\",\"showInAdvancedViewOnly\",\"unixHomeDirectory\",\"userAccountControl\",\"uSNChanged\",\"uSNCreated\",\"whenChanged\",\"whenCreated\"],\"AddedProperties\":[],\"RemovedProperties\":[],\"ModifiedProperties\":[],\"PropertyCount\":46,\"accountExpires\":0,\"adminCount\":1,\"audio\":[\"98\",\"97\"],\"badPasswordTime\":0,\"badPwdCount\":0,\"CanonicalName\":\"domain.local/Users/vagrant\",\"CN\":\"vagrant\",\"codePage\":1252,\"countryCode\":1,\"Created\":\"\\/Date(1568472722000)\\/\",\"createTimeStamp\":\"\\/Date(1568472722000)\\/\",\"Deleted\":null,\"Description\":\"vagrant\",\"DisplayName\":\"vagrant\",\"dSCorePropagationData\":[\"\\/Date(1568501833000)\\/\",\"\\/Date(1568472746000)\\/\",\"\\/Date(-11644473344000)\\/\"],\"instanceType\":4,\"isDeleted\":null,\"LastKnownParent\":null,\"lastLogoff\":0,\"lastLogon\":132174634828071852,\"lastLogonTimestamp\":132168556528845522,\"logonCount\":496,\"logonHours\":[255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255],\"memberOf\":[\"CN=Users,CN=Builtin,DC=domain,DC=local\",\"CN=Administrators,CN=Builtin,DC=domain,DC=local\"],\"Modified\":\"\\/Date(1575245238000)\\/\",\"modifyTimeStamp\":\"\\/Date(1575245238000)\\/\",\"nTSecurityDescriptor\":{\"AccessRightType\":\"System.DirectoryServices.ActiveDirectoryRights\",\"AccessRuleType\":\"System.DirectoryServices.ActiveDirectoryAccessRule\",\"AuditRuleType\":\"System.DirectoryServices.ActiveDirectoryAuditRule\",\"AreAccessRulesProtected\":true,\"AreAuditRulesProtected\":false,\"AreAccessRulesCanonical\":true,\"AreAuditRulesCanonical\":true},\"ObjectCategory\":\"CN=Person,CN=Schema,CN=Configuration,DC=domain,DC=local\",\"ObjectGUID\":\"8162ff17-3156-4aec-8c27-9a193de34a10\",\"objectSid\":{\"BinaryLength\":28,\"AccountDomainSid\":\"S-1-5-21-1956496868-4281279731-3754180002\",\"Value\":\"S-1-5-21-1956496868-4281279731-3754180002-1000\"},\"primaryGroupID\":513,\"ProtectedFromAccidentalDeletion\":false,\"pwdLastSet\":132129438715582354,\"sAMAccountName\":\"vagrant\",\"sAMAccountType\":805306368,\"sDRightsEffective\":15,\"showInAdvancedViewOnly\":false,\"unixHomeDirectory\":\"1\",\"userAccountControl\":66048,\"uSNChanged\":86105,\"uSNCreated\":8199,\"whenChanged\":\"\\/Date(1575245238000)\\/\",\"whenCreated\":\"\\/Date(1568472722000)\\/\"}"
}
```

And here is the output after the above change

```
ok: [dc01] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "filter": "sAMAccountName -eq 'vagrant'",
            "identity": null,
            "include_deleted_objects": false,
            "ldap_filter": null,
            "password": null,
            "properties": [
                "*"
            ],
            "search_base": null,
            "search_scope": null,
            "server": null,
            "username": null
        }
    },
    "objects": [
        {
            "CN": "vagrant",
            "CanonicalName": "domain.local/Users/vagrant",
            "Created": "2019-09-14T14:52:02.0000000+00:00",
            "Deleted": null,
            "Description": "vagrant",
            "DisplayName": "vagrant",
            "DistinguishedName": "CN=vagrant,CN=Users,DC=domain,DC=local",
            "LastKnownParent": null,
            "Modified": "2019-12-02T00:07:18.0000000+00:00",
            "Name": "vagrant",
            "ObjectCategory": "CN=Person,CN=Schema,CN=Configuration,DC=domain,DC=local",
            "ObjectClass": "user",
            "ObjectGUID": "8162ff17-3156-4aec-8c27-9a193de34a10",
            "ProtectedFromAccidentalDeletion": false,
            "accountExpires": 0,
            "adminCount": 1,
            "audio": [
                "Yg==",
                "YQ=="
            ],
            "badPasswordTime": 0,
            "badPwdCount": 0,
            "codePage": 1252,
            "countryCode": 1,
            "createTimeStamp": "2019-09-14T14:52:02.0000000+00:00",
            "dSCorePropagationData": [
                "2019-09-14T22:57:13.0000000+00:00",
                "2019-09-14T14:52:26.0000000+00:00",
                "1601-01-01T00:04:16.0000000+00:00"
            ],
            "instanceType": 4,
            "isDeleted": null,
            "lastLogoff": 0,
            "lastLogon": 132174634828071852,
            "lastLogonTimestamp": 132168556528845522,
            "logonCount": 496,
            "logonHours": "////////////////////////////",
            "memberOf": [
                "CN=Users,CN=Builtin,DC=domain,DC=local",
                "CN=Administrators,CN=Builtin,DC=domain,DC=local"
            ],
            "modifyTimeStamp": "2019-12-02T00:07:18.0000000+00:00",
            "nTSecurityDescriptor": "O:DAG:DAD:PAI(A;;LCRPLORC;;;AU)(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;SY)(A;;CCDCLCSWRPWPLOCRSDRCWDWO;;;BA)(A;;CCDCLCSWRPWPLOCRRCWDWO;;;DA)(A;;CCDCLCSWRPWPLOCRRCWDWO;;;S-1-5-21-1956496868-4281279731-3754180002-519)(OA;;CR;ab721a53-1e2f-11d0-9819-00aa0040529b;;WD)(OA;CI;RPWPCR;91e647de-d96f-4b70-9557-d63ff4f3ccd8;;PS)(OA;;CR;ab721a53-1e2f-11d0-9819-00aa0040529b;;PS)(OA;;RP;037088f8-0ae1-11d2-b422-00a0c968f939;4828cc14-1437-45bc-9b07-ad6f015e5f28;RU)(OA;;RP;037088f8-0ae1-11d2-b422-00a0c968f939;bf967aba-0de6-11d0-a285-00aa003049e2;RU)(OA;;RP;4c164200-20c0-11d0-a768-00aa006e0529;bf967aba-0de6-11d0-a285-00aa003049e2;RU)(OA;;RP;59ba2f42-79a2-11d0-9020-00c04fc2d3cf;4828cc14-1437-45bc-9b07-ad6f015e5f28;RU)(OA;;RP;bc0ac240-79a9-11d0-9020-00c04fc2d4cf;bf967aba-0de6-11d0-a285-00aa003049e2;RU)(OA;;RP;bc0ac240-79a9-11d0-9020-00c04fc2d4cf;4828cc14-1437-45bc-9b07-ad6f015e5f28;RU)(OA;;LCRPLORC;;4828cc14-1437-45bc-9b07-ad6f015e5f28;RU)(OA;;LCRPLORC;;bf967aba-0de6-11d0-a285-00aa003049e2;RU)(OA;;RP;59ba2f42-79a2-11d0-9020-00c04fc2d3cf;bf967aba-0de6-11d0-a285-00aa003049e2;RU)(OA;;RP;5f202010-79a5-11d0-9020-00c04fc2d4cf;4828cc14-1437-45bc-9b07-ad6f015e5f28;RU)(OA;;RP;4c164200-20c0-11d0-a768-00aa006e0529;4828cc14-1437-45bc-9b07-ad6f015e5f28;RU)(OA;;RP;46a9b11d-60ae-405a-b7e8-ff8a58d456d2;;S-1-5-32-560)(OA;;RPWP;6db69a1c-9422-11d1-aebd-0000f80367c1;;S-1-5-32-561)(OA;;RPWP;5805bc62-bdc9-4428-a5e2-856a0f4c185e;;S-1-5-32-561)(OA;;RPWP;bf967a7f-0de6-11d0-a285-00aa003049e2;;CA)",
            "objectSid": "S-1-5-21-1956496868-4281279731-3754180002-1000",
            "primaryGroupID": 513,
            "pwdLastSet": 132129438715582354,
            "sAMAccountName": "vagrant",
            "sAMAccountType": 805306368,
            "sDRightsEffective": 15,
            "showInAdvancedViewOnly": false,
            "uSNChanged": 86105,
            "uSNCreated": 8199,
            "unixHomeDirectory": "1",
            "userAccountControl": 66048,
            "whenChanged": "2019-12-02T00:07:18.0000000+00:00",
            "whenCreated": "2019-09-14T14:52:02.0000000+00:00"
        }
    ]
}
```